js/src/gc/RootMarking.cpp

```diff
@@ -48,7 +48,16 @@ using RootEnum = RootedValueMap::Enum;
 // ConcreteTraceable type are actually used at runtime.
 struct ConcreteTraceable {
   ConcreteTraceable() = delete;
-  void trace(JSTracer*) { MOZ_CRASH("This path is unreachable."); }
+#ifdef _MSC_VER
+  // There should be no callers to ConcreteTraceable::trace(). In clang/gcc, it
+  // is possible to ensure that using the "= delete" syntax, but MSVC is more
+  // aggressive about what is considered a caller.
+  void trace(JSTracer*) {
+    MOZ_ASSERT_UNREACHABLE();
+  }
+#else
+  void trace(JSTracer*) = delete;
+#endif
```

Comment:
```suggestion
void trace(JSTracer*) { MOZ_ASSERT_UNREACHABLE(); }
```

I'm pretty sure that `= delete` also leads to a compilation issue on GCC. Isn't it better to just use a default unreachable implementation?

Reply:
No it isn't better *IF* it can be avoided. A use of an `=delete` function cannot be compiled.
Maybe I don't know exactly what you mean by a "default unreachable implementation". This macro is a runtime failure, so that's not as good as a compile failure. But if there is actually a use of the function in some way that is implementation-dependent, then I agree we shouldn't leave it to chance.

I'm curious about how this ended up being a compiler-dependent issue, but not enough to block the PR.